File: leetcode/Valid_Sudoku.java
import java.util.*;

public class Solution {
    public boolean isValidSudoku(char[][] board) {
        for (int i = 0; i < 9; ++i) {
            boolean[] u = new boolean[9];
            for (int j = 0; j < 9; ++j)
                if (board[i][j] != '.') {
                    int x = board[i][j] - '1';
                    if (u[x]) return false;
                    u[x] = true;
                }
        }
        for (int j = 0; j < 9; ++j) {
            boolean[] u = new boolean[9];
            for (int i = 0; i < 9; ++i)
                if (board[i][j] != '.') {
                    int x = board[i][j] - '1';
                    if (u[x]) return false;
                    u[x] = true;
                }
        }
        for (int k = 0; k < 9; ++k) {
            boolean[] u = new boolean[9];
            for (int i = k / 3 * 3; i < (k / 3 + 1) * 3; ++i)
                for (int j = k % 3 * 3; j < (k % 3 + 1) * 3; ++j)
                    if (board[i][j] != '.') {
                        int x = board[i][j] - '1';
                        if (u[x]) return false;
                        u[x] = true;
                    }
        }
        return true;
    }
}

File: leetcode/Valid_Sudoku.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char> > &board) {
        // Row
        for (int i = 0; i < 9; ++i) {
            int bitmap = 0;
            for (int j = 0; j < 9; ++j) {
                if (board[i][j] == '.')
                    continue;
                int bit_pos = board[i][j] - '0';
                if ((bitmap & (1 << bit_pos)) != 0)
                    return false;
                else
                    bitmap |= (1 << bit_pos);
            }
        }
        // Column
        for (int j = 0; j < 9; ++j) {
            int bitmap = 0;
            for (int i = 0; i < 9; ++i) {
                if (board[i][j] == '.')
                    continue;
                int bit_pos = board[i][j] - '0';
                if ((bitmap & (1 << bit_pos)) != 0)
                    return false;
                else
                    bitmap |= (1 << bit_pos);
            }
        }
        // Cell
        for (int i = 0; i < 9; i += 3)
            for (int j = 0; j < 9; j += 3) {
                int bitmap = 0;
                for (int u = i; u < i + 3; ++u)
                    for (int v = j; v < j + 3; ++v) {
                        if (board[u][v] == '.')
                            continue;
                        int bit_pos = board[u][v] - '0';
                        if ((bitmap & (1 << bit_pos)) != 0)
                            return false;
                        else
                            bitmap |= (1 << bit_pos);
                    }
            }
        return true;
    }
};
